core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2378,28 +2378,47 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
   ): Stream[F2, O] =
     Stream.suspend {
       assert(maxFactor >= minFactor, "maxFactor should be greater or equal to minFactor")
-      val random = new scala.util.Random(seed)
-      def factor: Double = Math.abs(random.nextInt()) % (maxFactor - minFactor) + minFactor
 
-      def nextSize(sourceSize: Int): Int = (factor * sourceSize).toInt
+      underlying.uncons.flatMap {
+        case None => Pull.done
+        case Some((hd, tl)) =>
+          val random = new scala.util.Random(seed)
+          def factor: Double = Math.abs(random.nextInt()) % (maxFactor - minFactor) + minFactor
+
+          def nextSize(sourceSize: Int): Int = (factor * sourceSize).toInt
+
+          def go(
+              acc: Chunk[O],
+              sizeOpt: Int,
+              lastChunkSize: Int,
+              s: Pull[F2, O, Unit]
+          ): Pull[F2, O, Unit] = {
+
+            val size = if (sizeOpt > 0) sizeOpt else nextSize(lastChunkSize)
 
-      def go(acc: Chunk[O], sizeOpt: Int, s: Pull[F2, O, Unit]): Pull[F2, O, Unit] =
-        s.uncons.flatMap {
-          case Some((hd, tl)) =>
-            val size = if (sizeOpt > 0) sizeOpt else nextSize(hd.size)
             if (acc.size < size)
-              go(acc ++ hd, size, tl)
+              s.uncons.flatMap {
+                case None => Pull.output(acc)
+                case Some((hd, tl)) =>
```

Comment:
Nit: while reading the newest iteration of the code I had to do a double take, because there is now shading here (`hd` and `tl` used in the `go` function hide the variables with the same names from the higher scope of `underlying.uncons.flatMap`). Maybe it would be nice to avoid this for the sake of clarity. :slightly_smiling_face: